### DISCUSSION TITLE
Get tool_call_id from iterable response

### DISCUSSION THREAD

**Discussion by nitanmarcel** [567-labs/instructor#987] [Q&A]
I'm trying to use openai to execute consecutive commands in a shell, one after another based on the response of the previous command. For this I need to pass a new message with the role tool, and the result of the command. This requires the model to know which tool id to use. How can I retreive the tool_call_id?

```
                msg = {
                    "role": "tool",
                    "content": ANSI_REGEX.sub("", res),
                    "name": tool.name,
                    "tool_call_id": tool_call_id
                }
```

```python
client = from_openai(interpreter.openai_client, mode=Mode.TOOLS)

response = client.chat.completions.create(
    model=model,
    max_tokens=int(interpreter.env["llm.maxtokens"]),
    tools=tools,
    messages=interpreter.messages,
    tool_choice="auto",
    stream=True,
    temperature=float(interpreter.env["llm.temperature"]),
    response_model=Iterable[Response]
)
process_streaming_response(interpreter, response)

def process_streaming_response(interpreter, response):
    for chunk in response:
        tool_call_id = ?
````

**Comment by ReneFr81** [comment]
Based on my very limited experience, you can generate a unique ID (e.g., a UUID) manually and include it in both messages to keep track of tool calls. For example:

```json
[
  {"role": "system", "content": "You are an AI assistant..."},
  {"role": "user", "content": "user prompt"},
  {
    "role": "assistant", "content": null, "tool_calls": [
      {"id": "a4de1832-71bb-483d-9b58-85b97c11639e", "function": {"name": "Search...", "arguments": "{\"a\": \"1234\"}"}, "type": "function"}
    ]
  },
  {"role": "tool", "content": "{...}", "tool_call_id": "a4de1832-71bb-483d-9b58-85b97c11639e"}
]
```

Each time I initiate a tool call, I generate a UUID and assign it to both the `assistant` and `tool` messages. This approach allows me to track specific tool calls and their responses.

However, if I misunderstood your concern, feel free to elaborate! Let me know if this solution aligns with your requirements.

**Reply by nitanmarcel** [reply]
How do you track the tool call?

**Reply by ReneFr81** [reply]
What do you mean? I have a storage and message structure where I track the tool calls and responses. Everytime i call a tool I generate a uuid for that call and also use it for the result.
